### ISSUE TITLE
[TW-1572] Alternative approach to urgency inheritance

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#1598]
_Jens Erat on 2015-03-12T08:53:27Z says:_ 

 With 2.4, taskwarrior learned urgency inheritance such that the blocked task's urgency is added up.

This seems of rather limited use to me. In my opinion, a more reasonable approach would be to take the maximum of the blocked and blocking tasks' urgency, and assign this to the blocking task.

If I do have a blocked task which is really urgent (buying sunscreen) but blocked by a low-priority task (finding your rain suit, as it's horribly raining right now, but I wouldn't really care going outside otherwise), the low-priority task should be assigned the blocking task's urgency.

I'd expect implementing this would be a rather low-hanging fruit (a few lines around the current urgency inheritance code) and would tackle this with a patch, but would prefer some feedback before.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2015-03-12T08:53:27Z
Modified: 2017-01-16T17:27:32Z
```

**Comment by taskwarrior** [comment]
_Tomas Babej on 2015-03-12T10:29:26Z says:_ 

 Makes sense to me.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-03-13T00:45:36Z says:_ 

 I agree, setting urgency to the highest value in the chain makes a lot of sense.

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2015-03-13T08:11:01Z says:_ 

 My first reaction to this was "Hell yes! Why didn't I think of that back then?!". Then I re-read issue TW-296. It seems I really was blind to the obvious solution, as theorized there :)
1. 1, not inheriting certain values, still applies. You can't take "urgency_blocked" into account when inheriting urgency. Punishing child tasks for their parents' sins is still not socially acceptable.
1. 2, "long dependency chains lead to bloated urgency values" should be nicely solved by your proposal.
I also really like that it solves the horizontal counterpart of this problem as well: If a task is depended upon by several other low-urgency tasks, it shouldn't be more important just because it "unlocks" a lot of other tasks. Being able to work on low urgency tasks is not a valid reason to bump a parent task's priority up. If your idea works out, setting the "blocked" and "blocking" to 0 might be a good idea for people using urgency_inherit.
1. 3, efficiency, still applies. But it won't lead to more work than the currently implemented solution, so I think we're good.

Another thing that will come up with with your idea is that it is no longer, strictly speaking, an urgency coefficient. urgency_inherit will have to be either 0 or 1, since other values just don't make sense. But breaking with customs is a good thing in this case.

**Comment by taskwarrior** [comment]
_Jens Erat on 2015-03-13T08:22:36Z says:_ 

 Thank you for the comments.

I'm not sure whether there might be a use case for the coefficient (to be honest, I can't imagine one, it just does not seem to do the right thing for me). But different people will use Taskwarrior differently...
1. 1: Indeed. I guess I would've failed thinking of that in the first iteration (given the code I'd  be reusing isn't already taking care of that ;) ).
1. 2: Should be fine, but I'm still not sure whether a large number of dependant tasks shouldn't increase urgency. Or at least, they should be able to. Being _able_ to work at more tasks might also help burning down the queue, think of you're blocked of doing the most urgent tasks (time, location, whatever) and might burn down those instead.
1. 3: Yes, that shouldn't change anything.
1. 4: Yes, it would be a boolean then.

I will put together a first hack asap (I think it is a single changed line, anyway; after I had a short peek at the code). A long-term solution should have discussed, whether the current urgency inheritance model should be replaced or extended (thus, a new configuration variable introduced or the old one renamed/changed to a boolean).

**Comment by taskwarrior** [comment]
_Jens Erat on 2015-04-03T19:40:31Z says:_ 

 I started hacking on this, and realized the old inheritance system seems broken (it only adds up 1 for each urgency variable like priority, projects, ... but does not multiply with the factors; and it does not take UDAs and similar stuff into account).

I ended by adding up the urgency subtracted by being blocked to the result of urgency_c () instead and not adding inherited urgency, but taking the maximum: this seems to work quite reasonable.

I pushed a draft to my GitHub branch: https://github.com/JensErat/task/commit/2ecfd6f256b2651c725c194c06c40066b278a1c4

I didn't take care of spaces vs. tabs and similar and similar code style measures, don't even bother checking for these things; I will diff and clean up before providing a final commit.

Further things to do (feedback welcome):
*  -recursive inheritance- (done)
*  if we use the inherited urgency, add up the blocking coefficient anyway
*  also inherit due date (did I miss another attribute to inherit? maybe next?)
*  provide reasonable output for inherited urgency values when viewing a single task

**Comment by taskwarrior** [comment]
_Jens Erat on 2015-04-25T22:23:00Z says:_ 

 The newest draft includes all points mentioned in the last comment but the due date (which seems a little more complicate to do and subject to discussion).

I don't think I broke new tests but those who _should_ fail due to the changed behavior.

There's still stuff to do:
*  make inheritance configurable
*  either throw out the old adding up up urgency, or make it a complete and working alternative
*  add strings to the translation framework

Before I tackle those, I'd like some feedback on the general implementation. In my opinion, the new urgency calculation is much more reasonable. With my task database, I get reasonable urgency values for a blocked coefficient of 0.0 and a blocking coefficient of 1.0.

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2015-05-10T16:35:34Z says:_ 

 Thank you for working on this Jens.

For the record:
The previous implementation didn't multiply by the coefficients because that would have led to ballooning urgency values very quickly. If you inherit everything from your parent AND multiply it, you end up with huge urgency values, drowning out all tasks that are not in dependency chains. So what you call "broken" is just the way the previous implementation was designed to work. I'm pretty sure I explained all that in the linked issue.
But I already agreed that going for the maximum sounds like a way more reasonable approach, so all of this is beside the point. :)

I implemented "better urgency inheritance", i.e. your idea, in a local branch that I'm going to push to the central git repo once the 2.4.4 bugfix release is out.

As to your patch, I'll briefly describe why I didn't use it:
1.  It breaks tests. Not just ones testing old style inheritance, but unrelated functionality. You really want to make use of the test suite, it's your best friend during development.
1.  It discards urgency values less than 0.0
1.  It calls urgency_c() over and over again instead of calling urgency(), which caches urgency values.
1.  While showing inherited urgency in the info report is a good idea, I'm not entirely sure changing urgencyTerm() like that is the way to go. That said, I have not actually implemented an alternate version of this that I like any better yet, so this definitely open for discussion. As is everything else, of course.

**Comment by taskwarrior** [comment]
_Jens Erat on 2015-05-10T19:07:34Z says:_ 

 As I already pointed out, my patch was only "brainstorming in code" and not meant to be used as-is. Sometimes it's easier to describe things with a few lines of code compared with words...

1. If so, I missed that. I was pretty sure I diff'ed the test suite output: but I'm not familiar with the tests yet, definitely possible I simply misread something. If there were any output tests, I might also have ignored them.
2. I'm unsure what exactly you're pointing at. Consider a task A blocking task B (thus, B depending on A). A has negative urgency, B positive. I don't think A's urgency should influence B's, which is larger anyway.
3. I was unsure about that one, but there was something stopping me doing so (visibility of the function or simply sticking with other code I guessed doing similar stuff). Not sure about the reason any more, sadly I also did not elaborate in here. I don't know the Taskwarrior code very well yet, and struggle a lot with C++ anyway (never used it before).
4. I'd strongly propose to print _anything_. I would consider just keeping the table, but having the rows not sum up to the final urgency value being cruel against the user. ;) I see that adding a line of plain text is no elegant solution, but I couldn't come up with a better one (that's why I chose this one).

Adding to (3), another thing I noted was the repeated call of {code}dependencyGetBlocked{code}. I must admit I did not look to deeply into this function and am not completely sure what it does (Taskwarrior code lacks a bit of high-level comments on what's happening, and why), but considering this line I'm unsure whether we could save a few calls on it:

{code}
  // Calling dependencyGetBlocked is rather expensive.
  // It is called recursively for each dependency in the chain here.
{code}

Anyway: Could you share your current version somewhere? I'd love to have a look at it (and maybe even already switch to the implementation that might be used in future).

Thank you for pushing this forward!

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2015-05-11T08:48:14Z says:_ 

 1.  (failing tests) there are currently no failing tests, so "diffing" the test/run_all output shouldn't be necessary. What you're seeing is probably a result of missing the Perl JSON module, and missing "faketime". Once you have those on your system, all failures should be gone.
1.  (urgency lower than 0) you can actually see this in a failing test. Take a look. ;)
1.  (urgency_c() instead of urgency()) no need to defend your choice, I know perfectly well that the Taskwarrior source code is not exactly the easiest to jump into in certain places. I was listing reasons I didn't use your patch, not trying to attack you for it.
1.  (info command) I agree, we really should output _something_. Still not sure how to handle it.

[Here's the preliminary patch|http://wbsch.de/2015-05-11_urgency_inherit_unfinished.patch]. It doesn't include new tests yet, but it doesn't break old ones. Some more work and thorough testing is needed, as well as a good way to show that it's active in the info report.

**Comment by taskwarrior** [comment]
_Jens Erat on 2015-05-11T09:16:51Z says:_ 

 I have faktime already installed (handy in a lots of situations), I could indeed be missing the JSON module.

Don't worry -- I didn't feel attacked at all. Do you think we could safely reduce the number of {code}dependencyGetBlocked{code} calls? If this really does descend recursively, we should be able to remember some intermediate values or apply a similar strategy.

Did you consider other things we should inherit? If a blocked task has a shorter deadline than a blocking one, this would be a great candidate. "Waiting" tasks in the other way round could also be considered. I didn't really find a place to hack together such a change, finding the urgency code was much easier than digging deep into Taskwarrior's core of handling dependent tasks...

Regarding adjusting the output: another approach that would fit the current schema would be providing another line which equivs the difference between the current and maximum dependent task. But that's really hackyish, and I don't really like the idea. A better solution might be something like

{code}
    blocking            1 *    1 =      1
    tags              0.8 *    1 =    0.8
    age             0.066 *    2 =  0.132
    UDA priority.M      1 *  3.9 =    3.9
* -----
    Sum                             10.83
    Inherited Urgency               42.00
{code}

I don't think we could stroke-out the 10.83 on the command line, but we might print it in gray color or something like this. And maybe reference the task, something like "Inherited Urgency (#23)".

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2015-05-23T07:59:49Z says:_ 

 I have committed my working-but-not-optimal patch to the 2.4.5 branch in order to allow more people to play with it.

In order to test this the way it was meant to be used, adjust your taskrc like this:
```
urgency.inherit=on
urgency.blocked.coefficient=0.0
urgency.blocking.coefficient=0.0
report.next.filter="status:pending limit:page +READY"
```

Remaining problems / things to think about:
1.  The info report gives no indication of inherited urgency overruling regular urgency. This must be changed. The "how" of this is still open for discussion or, better yet, patches. I think I like your proposal of simply adding another line after the summed up urgency values.
1.  Inheriting urgency is currently a rather costly process. Precomputing (or caching) the necessary data and adding it as an attribute of a task for O(1) lookup would solve this problem, but lead to quite a bit of added complexity. Profiling and making a decision TBD.
1.  It might be prudent to keep the blocked coefficient, but prevent it from being inherited by re-adding the lost urgency when calling urgency_inherit(). This would hide, or at least move down, the blocked tasks in a dependency chain for reports that do not hide blocked tasks but add some complexity to the process. Thoughts welcome.
1.  The taskrc changes listed above will have to be made the default if urgency inheritance is to be a default option. I find that a blocking coefficient doesn't help my next report to be any better - one task unblocking a few more low urgency tasks does not warrant that blocking task being at the top of my "next" report if more urgent things need to be done. Thoughts welcome.
1.  Inheriting more task attributes like due date sounds reasonable, but should probably be discussed in a new issue.

**Comment by taskwarrior** [comment]
_Jens Erat on 2015-05-23T08:19:17Z says:_ 

 Regarding (3), I'm not sure if this is needed any more. If blocked tasks are not welcome in a report but printed, the report is broken. If you want to see them, being blocked does not actually change their urgency. On the other hand, (4) might well help you at getting stuff done: consider you're in a long flight, train ride or whatever, and cannot do everything (because you'd need internet access, be at home/in the office for the task, ...). Having as much as possible unblocked enables you to have anything reasonable to do, anyway. This might be a subjective need, but that's why coefficients are configurable.

I'll open a new issue for (5).

**Comment by taskwarrior** [comment]
_Jens Erat on 2015-12-30T22:32:18Z says:_ 

 There was an alternative patch implementing my proposal developed and merged, so I'd say this issue can be closed.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2017-01-16T17:27:32Z says:_ 

 This was addressed some time ago.

**Comment by pbeckingham** [comment]
[TW-1572_0001-Urgency-inheritance.patch.txt](https://github.com/GothenburgBitFactory/taskwarrior/files/1756390/TW-1572_0001-Urgency-inheritance.patch.txt)
